vmm/src/memory_manager.rs

```diff
@@ -228,8 +241,27 @@ impl MemoryManager {
             mem_end.unchecked_add(1)
         };
 
+        let mut virtiomem_region = None;
+        let mut virtiomem_resize = None;
         if let Some(size) = hotplug_size {
-            start_of_device_area = start_of_device_area.unchecked_add(size);
+            if hotplug_method == HotplugMethod::VirtioMem {
+                let start_addr = GuestAddress(
+                    (start_of_device_area.0 + vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE - 1)
+                        / vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE
+                        * vm_virtio::VIRTIO_MEM_DEFAULT_BLOCK_SIZE,
+                );
+                virtiomem_region = Some(MemoryManager::create_ram_region(
```

Comment:
Please educate me: I am trying to figure out why we have to create the region and map it to guest at the start of VM Why we can't do it later when resize is called like ACPI based hotplug mem does? Thanks

Reply:
The region is allocated in the initialization phase because I haven't implemented resizable allocations yet.  So all the virtio-memory need to be allocated at the same time.  Sometimes the allocation fails because the size exceeds the system memory limit.  I think fail in the begin better than fail when VM is running.

And I haven't implemented resizable allocations because I think it is better to implement this function into vm-memory.

Reply:
The way `virtio-mem` is designed and as long as you can start the VM, that means you can later hotplug more memory (if you stay in the pre-defined bounds). So I agree with @teawater that it's important to make sure the memory can be allocated upfront in the context of virtio-mem.

Reply:
I still don't think we should "allocate" all the memory when starting of VM.
Here, memory allocation for VM includes 1. Creating GuestRegionMmap and 2.Call KVM to set_user_memeory_region.
Let's have a look at how GuestRegionMmap is implemented in vm-memory

impl MmapRegion {
    /// Creates a shared anonymous mapping of `size` bytes.
    ///
    /// # Arguments
    /// * `size` - The size of the memory region in bytes.
    pub fn new(size: usize) -> Result<Self> {
        Self::build(
            None,
            size,
            libc::PROT_READ | libc::PROT_WRITE,
            libc::MAP_ANONYMOUS | libc::MAP_NORESERVE | libc::MAP_PRIVATE,
        )
    }
Here, the flags MAP_ANONYMOUS | libc::MAP_NORESERVE | libc::MAP_PRIVATE means it only allocates the Host Virtual Address, and the Host Physical pages are not populated.
yet. Since we have enough Virtual memory space on 64bit machine. I don't think it may fail even allocating them later. Besides, in this situation only after the corresponding Guest Memory  Blocks are plugged, and the memory has been written, Host Physical pages then will be populated by Host kernel in Page Fault Exception. Only at this time, it may fail if no available physical memory pages.
So Creating GuestRegionMmap on the starting of VM  actually can't reserve the Host physical memory for Virtion-mem to use. Instead, Creating GuestRegionMmap and Call KVM to set_user_memeory_region only increases the latency of the starting of VM.
We may populate Host physical memory using flag MAP_POPULATE. But I don't think that is the design goal for memory hotplug.